kernel/src/drivers/bga.c

```
@@ -0,0 +1,137 @@
+#define BGA_VERSION         0xB0C5
+
+static int bga_detect(void)
+{
+    bga_write(BGA_ID, BGA_VERSION);
+    return (bga_read(BGA_ID) == BGA_VERSION);
+}
+
+/* PCI config read (simplified: only for QEMU VGA at bus 0 dev 2 func 0) */
+static uint32_t pci_read32(uint8_t bus, uint8_t dev, uint8_t func, uint8_t offset)
+{
+    uint32_t addr = (1 << 31)                   /* enable bit */
```

Cppcheck (2.21.0)

[error] Shifting signed 32-bit value by 31 bits is undefined behaviour

(shiftTooManyBitsSigned)